### ISSUE TITLE
False positive `the final variable can only be set once`

### ISSUE THREAD

**Issue by AlexanderFarkas** [dart-lang/sdk#53413]
```dart
 Future<void> _process(Future<T> Function() execute) async {
    final Future<T> future;
    try {
      future = execute();
    } catch (e, s) {
      future = Future.error(e, s);
    }
 }
```
I doesn't compile with error: `The final variable 'future' can only be set once`.

Dart SDK version: 3.1.0 (stable) on "macos_arm64"

**Comment by lrhn** [comment]
I believe the floor analysis is not clever enough to know that no errors can be thrown after the assignment to `future`.

If there had been any code after the assignment, which could potentially throw (so any code at all), then it would be a potential second assignment in the catch.

Generally, a catch clause assumes it can be reached from any place in the try block. That means that a variable is potentially assigned if it's assigned anywhere in the try block. Which it is, right after the assignment.

It's possible that we could exclude the position right after the last operation that can possibly throw, because if we get there, it didn't throw. It's just that there are so very few operations that cannot throw, with reading and writing local variables being some of the exceptions.

**Comment by stereotype441** [comment]
> That is: This is _likely_ working as specified. If we want to change it, it's a language change. It would probably be a non-breaking change. A variable going from potentially assigned to unassigned will not prevent more reads (unless it's a late local variable, which is being read while only potentially assigned) and will allow more writes. (_Someone_ who understands the of assingment analysis definition around `try`/`catch` better may be able to confirm this. @stereotype441?)

Yes, flow analysis was designed this way on purpose. One of the principles of its design is that its behaviour should be stable if the user adds innocuous code (for example calls to logging functions), since people often do such things as a temporary measure during debugging. So, in the example above, we want to allow for the possibility that someone comes along later and does this:

```dart
Future<void> _process(Future<T> Function() execute) async {
  final Future<T> future;
  try {
    future = execute();
    log('execution succeeded');
  } catch (e, s) {
    future = Future.error(e, s);
  }
}
```

Since the function `log` might in principle do anything, it's possible that it could throw an exception, which would cause the local variable `future` to get set twice. We don't want to create a situation where the user can't add a logging statement without creating other errors.

(Also, flow analysis was substantially easier to implement this way, because rather than carefully consider exactly which statements and expressions could possibly throw exceptions, it can simply scan the `try` block to see which variables it writes to, and consider all those variables to be "maybe written to, maybe not" in the `catch` block.)

**Comment by AlexanderFarkas** [comment]
Sounds reasonable to me
